SILFunctionConventions

atrick opened the pull request:
I've extracted these *mostly* NFC changes from a branch where I'm working on representing opaque SIL types as SSA values. I'd like to get this upstreamed ASAP now that 3.1 has branched. Merging it has been a big problem over the past few weeks.

In hindsight I can see how this could further be split into more commits, but it difficult enough just to separate it from other changes related to argument conventions. Reviewers can just look at AST/Types.h and SIL/SILType.h to see the new API without scanning the whole diff.

I initially implemented a less intrusive approach: add a "fake" paramter/result convention for opaque types. That approach was wrong in principle because parameter/result conventions should only reflect ABI/reabstraction considerations. That approach also had serious problems in practice. It became very hard to separate SIL type lowering from IRGen (not to be confused with formal type lowering). It was very easy to misuse (bug prone), and led to nonuniformity of the SIL API across compilation stages. A priority for me is to establish uniform idioms for working with SIL regardless of the compilation stage.

The approach in the PR is extremely intrusive but establishes uniformity across stages and clear lines of responsiblity. SILFunctionTypes and parameter/result conventions are part of the formal type system, even after formal type lowering. SILTypes are separate and determined by the current SIL module. SIL function arguments have their own SILType-level conventions based on the current module configuration and stage.

atrick committed:
[SILType] SILFunctionConventions API.

Separate formal lowered types from SIL types.
The SIL type of an argument will depend on the SIL module's conventions.
The module conventions are determined by the SIL stage and LangOpts.

Almost NFC, but specialized manglings are broken incidentally as a result of
fixes to the way passes handle book-keeping of aruments. The mangler is fixed in
the subsequent commit.

Otherwise, NFC is intended, but quite possible do to rewriting the logic in many
places.

atrick committed:
[NFC] In AllocBoxToStack, Use "ArgIndex" for argument indices.

Names are important.

atrick committed:
[NFC] Specializationmangler ArgNo -> OrigArgIdx.

Names matter. When using an unsigned int to index arguments, always make it
clear what the index refers to. It is a particularly confusing in this code because:
- mangling should not care about argument indices at all, only the function type should matter.
- argument indices for a given function type may be different depending on the SIL stage.
- these indices are actually a contract between the client code and the mangler.
- the specialized function's argument indices are different than the original indices!

This issue was hiding bugs in the mangler. The bug fixes will be in a separate PR.

atrick committed:
Fix the specialized mangler.

Pass correct argument indices to the mangler for function specialization. This
has probably always been broken. The demangler doesn't support these manglings
anyway. It doesn't seem to have mattered in practice yet (aside from the
confusion of devs) like me. It will matter once we begin exposing public
specializations.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test.

gottesmm commented:
Andy, before you commit, do you think you can rebase so we do not have the merge related commits?

gottesmm commented on lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:
```diff
@@ -258,10 +258,10 @@ static size_t getParameterIndexForOperand(Operand *O) {
     auto ArgCount = PartialApply->getArguments().size();
     assert(ArgCount <= FnType->getParameters().size());
-    ParamIndex += (FnType->getNumSILArguments() - ArgCount);
+    ArgIndex += (FnType->getNumSILArguments() - ArgCount);
```
I am pretty sure the (...) is unneeded, no?

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -2713,7 +2711,7 @@ enum class ParameterConvention {
 };
 
-inline bool isIndirectParameter(ParameterConvention conv) {
+inline bool isIndirectFormalParameter(ParameterConvention conv) {
```
Is it possible to make ParameterConvention a struct enum? Then this could be a method!

I am fine with this being in a different commit/PR.

atrick replied:
Sure. Despite the appearance, I was trying *not* to do unnecessary cleanup in this PR.

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -2878,7 +2879,8 @@ enum class ResultConvention {
 
-inline bool isIndirectResult(ResultConvention convention) {
+// Is this result passed indirectly for the purpos of reabstraction.
+inline bool isIndirectFormalResult(ResultConvention convention) {
```
This is something else that could become a method with a struct enum.

atrick replied:
It could be.

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -3094,81 +3097,63 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
     return *reinterpret_cast<CanType*>(ptr);
   }
 
+  CanType &getMutableAllResultsCache() const {
```
I noticed that above you eliminated the All from getMutableAllResults(). Why leave it here?

atrick replied:
Because `getAllResults` did *not* return all results. It was misnamed. `getMutableAllResultsCache` does cache all results.

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -3094,81 +3097,63 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
 
+  CanType &getMutableAllResultsCache() const {
+    assert(hasResultCache());
+    auto ptr = const_cast<SILFunctionType*>(this)->getEndOfNormalResults()
```
Can you change `auto ptr` => `auto *ptr` and put a space in between the '*' and SILFunctionType.

atrick replied:
Agreed and done. (again, I was avoiding cleanup here)

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -3178,31 +3163,16 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
   ///   - a single indirect result and no direct results.
-  SILType getCSemanticResult();
+  ///
+  /// If the result is formally indirect, returns the empty tuple.
```
Shouldn't returns be return here?

atrick replied:
That grammar seems preferred in this file. Done.

gottesmm commented on include/swift/AST/Types.h:
```diff
@@ -3213,55 +3183,99 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
-    return NumIndirectResults != 0;
+  };
+  using IndirectFormalResultIter
+  = llvm::filter_iterator<const SILResultInfo*, IndirectFormalResultFilter>;
```
'const SILResultInfo*' => 'const SILResultInfo *'

atrick replied:
clang-format caught all of these. That's a new convention to me.

gottesmm commented on include/swift/SIL/SILArgument.h:
```diff
@@ -116,7 +117,7 @@ class SILArgument : public ValueBase {
 
   // A special constructor, only intended for use in
-  // SILBasicBlock::replacePHIArg.
+  // SILBasicBlock::replacePHIArg and replaceFunctionArg.
```
I think I removed replaceFunctionArg. So this comment I believe is incorrect.

atrick replied:
I know. That annoyed me since I need it ;) It will be used in the next PR.

gottesmm commented on include/swift/SIL/SILArgument.h:
```diff
@@ -234,6 +229,12 @@ class SILFunctionArgument : public SILArgument {
+
+  // A special constructor, only intended for use in
+  // SILBasicBlock::replaceFunctionArg.
+  explicit SILFunctionArgument(SILType Ty, ValueOwnershipKind OwnershipKind,
```
Again, I think I got rid of SILBasicBlock::replaceFunctionArg.

gottesmm reviewed (commented):
I did a small pass through of the first 2 commits and parts of the 3rd.

I would suggest running clang-format over the commits. Also, I think I eliminated SILBasicBlock::replaceFunctionArg. I think you picked up some code related to that when you were merging. I am going to take another dive through it in a bit.

atrick commented:
@gottesmm I rebase constantly: every time I update and retest, and never merge locally. But github merges so it can kick off the CI jobs. Then if you rebase again in the middle of a PR you lose the testing and any reviews others were working on.

atrick commented:
I'm waiting for these tests before rebasing and applying clang-format.

atrick commented:
@jckarter Any concerns? Want some time to look this over?

atrick commented:
@swift-ci test.

atrick commented:
Note: I used `git-clang-format --style "{BasedOnStyle: LLVM, BreakBeforeBinaryOperators: All}"`. Which does:

```
Type var
  = value;
```

I'd be happy to use `BreakBeforeBinaryOperators: BOS_NonAssignment` instead. I don't know what's preferred.

atrick commented:
For the record, this passed full swift-ci testing. I may do some cleanup and just re-smoke-test.

gottesmm commented on include/swift/SIL/SILType.h:
```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
+  friend SILResultInfo;
+  friend SILFunctionConventions;
+
+  static bool isIndirectSILParam(SILParameterInfo param, bool loweredAddresses);
```
Why do we need these static methods? When will we not have access to SILModuleConventions and not have access to lowered addresses?

atrick replied:
The static methods are used by the implementations of SILParameterInfo and SILResultInfo. When we want formal types, we just ask for lowered addresses without any SIL context.

gottesmm commented on include/swift/SIL/SILType.h:
```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
+/// conventions.
+class SILFunctionConventions {
+public:
+  SILModuleConventions silConv;
```
Why make these public? Why not just make a getter?

atrick replied:
The "conventions" object can be on-demand. A getter would logically construct
a new one each time. That's not what I want for most uses.

More importantly, `silConv` is an interface to conceptually global
context. Like `M`, `F`, `IGM`, `SGM`, etc. It's important to
standardize abbreviations for these. They should be accessed using
common idioms throughout the compiler.

Defining getters for global context is superfluous. It is not part of
the class's implementation detail. The only reason we attach these
contextual artifacts to most of our classes is because of a decision
not to use TLS within the compiler. That decision makes exploratory
programming hard, and makes it difficult to establish uniform
idioms. We can improve the situation by standarizing on simple
abbreviated names for contextual state and making them a public member
of every class that accesses them.

I'm taking the time to explain this not because of some theoretical
concern but because lack of standard idioms is a real problem for me
when working across different areas of our code base. Most of my
cognitive load goes to figuring out how the same familiar things are
accessed differently in each corner of the code.

gottesmm commented on include/swift/SIL/SILType.h:
```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
+    return makeIteratorRange(filter.begin(), filter.end());
+  }
+
+  struct SILResultTypeFunc {
```
Should uses ever see this? Maybe hide it in a detail struct? Not wedded to this, just a thought.

atrick replied:
Yep. But `private` works fine in a class if you bother to use it.

atrick commented:
@gottesmm @jckarter 
Here's a summary of the salient changes. I wish I could have presented this better:

Types.h and ASTContext.cpp
- Remove the direct association between formal conventions, formal
  types, and SIL types.
- Change the representation of SILFunctionType since we now have
  SIL-level abstractions over indirect vs direct results.

SILType.h, SILType.cpp, and SILFunctionType.cpp
- Add the SILFunctionConventions API to map between formal conventions
  and SIL types.

SILInstruction
- ApplySite::getCalleeArgIndexOfFirstAppliedArg().

SILOwnershipVerifier.cpp
- Quickly merged.

Other less-than trivial files:
- SILVerifier.cpp
- CapturePromotion.cpp
- EagerSpecializer.cpp
- Devirtualize.cpp
- GenericCloner.cpp
- Generics.cpp

Most of the other changes are formulaic adjustments to the API, convenience, and naming things for clarity.

But although it looks formulaic, it was actually very hard for me to determine in all these cases where we cared about formal parameters and results vs. SIL arguments. The mappings between results, arguments, and applied arguments was also pretty confusing in some cases.

gottesmm commented on include/swift/SIL/SILType.h:
```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
+      return funcTy->getIndirectFormalResults();
+
+    auto filter = llvm::make_filter_range(
+        makeIteratorRange((const SILResultInfo *)0, (const SILResultInfo *)0),
```
nullptr

atrick replied:
Oops. Old habit.

gottesmm reviewed (commented):
All of this looks reasonable to me. Some small additional changes. I will only really know how I feel about it when using it. If there are issues, we can iterate.

atrick commented:
@jckarter & @rjmccall: I don't need a code review. @gottesmm already did that (thanks).

I do want to make sure there isn't a serious concern with the terminology that I'm introducing (SILType.h and Types.h).

See the swift-dev Proposal: Opaque SIL values

I need to clearly distinguish the layers in the type system to prevent
misuse. In particular, I need to clarify when asking for a parameter's
"SIL" type whether I'm asking about the type that represents the
formal parameter, or the type that represents the SIL argument
value. In the first case (formal paramter) I say that we're asking for
a lowered formal type. That makes it clear that the query should be
used when we care about language requirements (e.g. reabstraction),
not the actual representation of SIL values. Similarly, code should
not simply rely on the SILType when dealing with parameter conventions.

Here's how I sort of fit that into my worldview:

- Formal types: The AST types directly exposed by the language.

- Canonical formal types: desugared.

- Lowered formal types: Canonical types in the ASTContext that
  represent language-level calling conventions and ABI requirements in
  terms of Swift types. Namely, SILFunctionType. Caller/callee lowered
  types need to match. They have been lowered such that it is valid to
  associate them with SIL values, but they exist independent of a
  SILModule.

- SIL types: The actual type of any SIL value. These merely wrap an
  address vs. object distinction around a lowered formal type. They
  can't exist without a SILModule, and reflect the SILModule's
  conventions. Mapping lowered formal types to SIL types is specific
  to the current SIL stage.

- SIL types with lowered addresses: Represent the ABI requirements for
  indirection and storage of SIL objects. All SIL types in the
  "lowered" stage have lowered addresses.

- LLVM types: Represent the ABI requirements in terms of C types. This
  could introduce additional indirection, but I'd like to handle most
  if not all of that in SIL address lowering.

jckarter commented:
What you call "lowered formal type", I've been thinking of as just "lowered type". Otherwise, that seems like a reasonable breakdown of the distinctions.

atrick commented:
@jckarter I think "lowered AST" type would be a more correct term. I mainly care about the SILFunctionType-related API surface though. `isFormallyIndirect` vs. `isSILIndirect` seemed the most intuitive to me.

jckarter commented:
To me, "AST type" also implies "formal type". If the intent is to distinguish the storage type from the address/object category of the type, what do you think of "storage type"?

atrick commented:
I'd be happy talking about SIL storage types vs. SIL value types. I'll try to write descriptions/comments in those terms.

I still think of our `ParameterConvention` as a formal convention vs. a SIL conventions. So what do I do with SILFunctionType? In this PR, it has methods:

- getIndirectFormalParameters()
- getIndirectFormalResults()
etc.

Wrapping the SILFunctionType around a SILFunctionConventions adds SIL-level conventions:
- getIndirectSILParameters()
- getIndirectSILResults()

I think that clearly communicates the level of abstraction. As opposed to for example:
- getIndirectlyStoredParameters()
vs.
- getIndirectSILParameters()

atrick committed:
Fix terminology. There's no such thing as a "formal SIL type".

Formal types are defined by the language's type system. SIL types are
lowered. They are no longer part of that type system.

The important distinction here is between the SIL storage type and the SIL value
type. To make this distinction clear, I refer to the SILFunctionTypes "formal"
conventions. These conventions dictate the SIL storage type but *not* the SIL
value type. I call them "formal" conventions because they are an immutable
characteristic of the function's type and made explicit via qualifiers on the
function type's parameters and results. This is in contrast to to SIL
conventions which depend on the SIL stage, and in the short term whether the
opaque values flag is enabled.

atrick commented:
I fixed up the terminology because there's no such thing as a formal SIL type. See the latest commit.

I reran git-clang-format --style "{BasedOnStyle: LLVM, BreakBeforeBinaryOperators: NonAssignment}"

atrick commented:
@swift-ci smoke test and merge.

atrick commented:
@swift-ci smoke test and merge.
